modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm

```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
+          push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+          undef $pending_ancestral_species;
+        }
         ## DEBUG info
-        # print "NODE LINK!\n";
-        # for (my $i = 0; $i<@$species_order; $i++) {
-        #   if ($i == $species_counter) {
-        #     print $species_order->[$i]->{genome_db}->name, "***\n";
-        #   } else {
-        #     print $species_order->[$i]->{genome_db}->name, "\n";
-        #   }
-        # }
-
-      ## 2. Force insertions in a math to the next right_node is expected
-      } elsif (defined($species_right_node_id) and exists($existing_node_ids->{$species_right_node_id})) {
-        splice(@$species_order, $species_counter, 0, {
-            genome_db => $this_genome_db,
-            right_node_id => $this_right_node_id,
-            genomic_align_ids => [@$these_genomic_align_ids],
-          });
-        ## DEBUG info
-        # print "FORCE INSERT!\n";
-        # for (my $i = 0; $i<@$species_order; $i++) {
-        #   if ($i == $species_counter) {
-        #     print $species_order->[$i]->{genome_db}->name, "***\n";
-        #   } else {
-        #     print $species_order->[$i]->{genome_db}->name, "\n";
-        #   }
-        # }
-
-      ## 3. If there is no info about right node or this points to a node not found in next tree,
+        # _debug_info("NODE LINK", $species_order, $species_counter);
+
+      ## 2. If there is no info about right node or this points to a node not found in next tree,
       ## rely on the species name
-      } elsif ($this_genome_db->name eq $species_genome_db->name
-                and (!defined($species_right_node_id) or
-                    !defined($existing_node_ids->{$species_right_node_id}))
-          ) {
+      } elsif (defined $species_genome_db and ($species_genome_db->name eq $this_genome_db->name)) {
+        my $debug_msg = "MATCH";
+        if ($this_genome_db->name eq "ancestral_sequences") {
+          $pending_ancestral_species = {
+            genome_db         => $this_genome_db,
+            right_node_id     => $this_right_node_id,
+            genomic_align_ids => [ @$these_genomic_align_ids ],
+          };
+          $debug_msg = "TENTATIVE " . $debug_msg;
+        } else {
+          $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
+          push (@{$species_order->[$species_counter]->{genomic_align_ids}}, @$these_genomic_align_ids);
+          if (defined $pending_ancestral_species) {
+            my $prev_species_index = $species_counter - 1;
+            $species_order->[$prev_species_index]->{right_node_id} = $pending_ancestral_species->{right_node_id};
+            push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+            undef $pending_ancestral_species;
```

Comment:
It's hard to follow because of varying indentations (I know, not your fault) - just want to check and take your word for it that `$pending_ancestral_sequences` is not trapped in an always `undef` state! Also to double check that `undef $pending_ancestral_sequences` does as you are intending it to do. `undef` like this seems to work in unexpected ways!

Reply:
In all my testing it worked as expected: when the species is an ancestral sequence, unless the `right_node_id` is found on the tree, it is stored in `$pending_ancestral_sequences` to see if the next species in the tree matched the next one in `$species_order` (to aggregate ancestral sequences accordingly). I also feared `undef` way delete the content (so I was cautious and tested it thoroughly) but since it is a reference, it seems it is dereferencing it only (and since the data has been copied somewhere else, it is not deleted).